cli/lsp/diagnostics.rs

```diff
@@ -54,61 +56,145 @@ pub struct DiagnosticServerUpdateMessage {
-type DiagnosticsByVersionMap = HashMap<Option<i32>, Vec<lsp::Diagnostic>>;
+struct DiagnosticRecord {
+  pub specifier: ModuleSpecifier,
+  pub versioned: VersionedDiagnostics,
+}
+
+#[derive(Clone, Default, Debug)]
+struct VersionedDiagnostics {
+  pub version: Option<i32>,
+  pub diagnostics: Vec<lsp::Diagnostic>,
+}
 
-#[derive(Clone)]
+type DiagnosticVec = Vec<DiagnosticRecord>;
+
+#[derive(Debug, Hash, PartialEq, Eq, Copy, Clone)]
+pub enum DiagnosticSource {
+  Deno,
+  Lint,
+  Ts,
+}
+
+impl DiagnosticSource {
+  pub fn as_lsp_source(&self) -> &'static str {
+    match self {
+      Self::Deno => "deno",
+      Self::Lint => "deno-lint",
+      Self::Ts => "deno-ts",
+    }
+  }
+}
+
+type DiagnosticsBySource = HashMap<DiagnosticSource, VersionedDiagnostics>;
+
+#[derive(Debug)]
 struct DiagnosticsPublisher {
   client: Client,
-  all_diagnostics:
-    Arc<Mutex<HashMap<ModuleSpecifier, DiagnosticsByVersionMap>>>,
+  diagnostics_by_specifier:
+    Mutex<HashMap<ModuleSpecifier, DiagnosticsBySource>>,
 }
 
 impl DiagnosticsPublisher {
   pub fn new(client: Client) -> Self {
     Self {
       client,
-      all_diagnostics: Default::default(),
+      diagnostics_by_specifier: Default::default(),
     }
   }
 
   pub async fn publish(
     &self,
+    source: DiagnosticSource,
     diagnostics: DiagnosticVec,
     token: &CancellationToken,
-  ) {
-    let mut all_diagnostics = self.all_diagnostics.lock().await;
-    for (specifier, version, diagnostics) in diagnostics {
+  ) -> usize {
+    let mut diagnostics_by_specifier =
+      self.diagnostics_by_specifier.lock().await;
+    let mut seen_specifiers = HashSet::with_capacity(diagnostics.len());
+    let mut messages_sent = 0;
+
+    for record in diagnostics {
       if token.is_cancelled() {
-        return;
+        return messages_sent;
       }
 
-      // the versions of all the published diagnostics should be the same, but just
-      // in case they're not keep track of that
-      let diagnostics_by_version =
-        all_diagnostics.entry(specifier.clone()).or_default();
-      let version_diagnostics =
-        diagnostics_by_version.entry(version).or_default();
-      version_diagnostics.extend(diagnostics);
+      seen_specifiers.insert(record.specifier.clone());
+
+      let diagnostics_by_source = diagnostics_by_specifier
+        .entry(record.specifier.clone())
+        .or_default();
+      let version = record.versioned.version;
+      let source_diagnostics = diagnostics_by_source.entry(source).or_default();
+      *source_diagnostics = record.versioned;
+
+      // DO NOT filter these by version. We want to display even out
+      // of date diagnostics in order to prevent flickering. The user's
+      // lsp client will eventually catch up.
+      let all_specifier_diagnostics = diagnostics_by_source
+        .values()
+        .flat_map(|d| &d.diagnostics)
+        .cloned()
+        .collect::<Vec<_>>();
```

Comment:
The main fix is here. We store the diagnostics by specifier then source (deno, lint, ts) and then don't filter on version.